cupyx/scipy/interpolate/_polyint.py

```diff
@@ -0,0 +1,323 @@
+            ny = len(self._y_extra_shape)
+            s = (list(range(nx, nx + self._y_axis))
+                 + list(range(nx)) + list(range(nx + self._y_axis, nx + ny)))
+            y = y.transpose(s)
+        return y
+
+    def _reshape_yi(self, yi, check=False):
+        """
+        Reshape the updated yi to a 1-D array
+        """
+        yi = cupy.moveaxis(cupy.asarray(yi), self._y_axis, 0)
+        if check and yi.shape[1:] != self._y_extra_shape:
+            ok_shape = "%r + (N,) + %r" % (self._y_extra_shape[-self._y_axis:],
+                                           self._y_extra_shape[:-self._y_axis])
+            raise ValueError("Data must be of shape %s" % ok_shape)
+        return yi.reshape((yi.shape[0], -1))
+
+    def _set_yi(self, yi, xi=None, axis=None):
+        if axis is None:
+            axis = self._y_axis
+        if axis is None:
+            raise ValueError("no interpolation axis specified")
+
+        shape = yi.shape
+        if shape == ():
+            shape = (1,)
+        if xi is not None and shape[axis] != len(xi):
+            raise ValueError("x and y arrays must be equal in length along "
+                             "interpolation axis.")
+
+        self._y_axis = (axis % yi.ndim)
+        self._y_extra_shape = yi.shape[:self._y_axis]+yi.shape[self._y_axis+1:]
+        self.dtype = None
+        self._set_dtype(yi.dtype)
+
+    def _set_dtype(self, dtype, union=False):
+        if cupy.issubdtype(dtype, cupy.complexfloating) \
+                or cupy.issubdtype(self.dtype, cupy.complexfloating):
+            self.dtype = cupy.complex_
+        else:
+            if not union or self.dtype != cupy.complex_:
+                self.dtype = cupy.float_
+
+
+class BarycentricInterpolator(_Interpolator1D):
+    """The interpolating polynomial for a set of points.
+
+    Constructs a polynomial that passes through a given set of points.
+    Allows evaluation of the polynomial, efficient changing of the y
+    values to be interpolated, and updating by adding more x values.
+    For reasons of numerical stability, this function does not compute
+    the coefficients of the polynomial.
+
+    The value `yi` need to be provided before the function is
+    evaluated, but none of the preprocessing depends on them,
+    so rapid updates are possible.
+
+    Parameters
+    ----------
+    xi : array-like
+        1-D array of x-coordinates of the points the polynomial should
+        pass through
+    yi : array-lie, optional
+        The y-coordinates of the points the polynomial should pass through.
+        If None, the y values will be supplied later via the `set_y` method
+    axis : int, optional
+        Axis in the yi array corresponding to the x-coordinate values
+
+    See Also
+    --------
+    scipy.interpolate.BarycentricInterpolator
+
+    """
+
+    def __init__(self, xi, yi=None, axis=0):
+        _Interpolator1D.__init__(self, xi, yi, axis)
+
+        self.xi = cupy.asfarray(xi)
```

Comment:
Let's use `astype` as `cupy.asfarray` internally uses `cupy.array`, which implicitly converts numpy array to cupy array.